oauth_samples/go/oauth.go

```
@@ -0,0 +1,155 @@
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/google/uuid"
+	"golang.org/x/oauth2"
+)
+
+const (
+	authorizeURL = "https://twist.com/oauth/authorize"
+	tokenURL     = "https://twist.com/oauth/access_token"
+
+	// callback url specified when the application was defined
+	// eg. https://dev.twist.test:4567/v3/
+	callbackURL = "<<enter the callback url for your application>>"
+
+	// e.g. https://api.twist.com/api/v3/workspaces/get
+	apiURL = "<<enter the url of the api endpoint you wish to access>>"
+)
+
+var (
+	// Should be a comma separated list, see https://developer.twist.com/v3/#scopes for a full list
+	// user:read,workspaces:read
+	scopes = "<<enter the scopes your application requires>>"
+	state  = uuid.New().String()
+
+	// client (application) credentials - located at https://twist.com/integrations/build, then go to your integration
+	clientID     = os.Getenv("twist_clientID")
+	clientSecret = os.Getenv("twist_clientSecret")
+)
+
+func main() {
+	if err := requestAuthorization(); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func requestAuthorization() error {
+	reader := bufio.NewReader(os.Stdin)
+
+	// Step 1 - simulate a request from a browser on the authorize_url -
+	// will return an authorization code after the user is prompted for credentials.
+	vals := make(url.Values)
+	vals.Set("response_type", "code")
+	vals.Set("client_id", clientID)
+	vals.Set("redirect_url", callbackURL)
+	vals.Set("scope", scopes)
+	vals.Set("state", state)
+
+	fmt.Println("Go to the following url on a browser and enter the code from the returned url")
+	fmt.Println(authorizeURL + "?" + vals.Encode())
+	fmt.Printf("-> ")
+
+	// Read the authorization code
+	text, readerError := reader.ReadString('\n')
+
+	if readerError != nil {
+		fmt.Println("Erorr reading the code you entered")
+		return readerError
```

Instead of printing a message followed by error return, it's better to just return an annotated error here and let the `main` function print non-nil error and fail the program:

```go
return fmt.Errorf("Error reading the code you entered: %v", readerError)
```